python/cudf/cudf/core/index.py

```diff
@@ -2158,6 +2164,27 @@ def __init__(
 
         super().__init__(data, **kwargs)
 
+        if self._freq is not None:
+            unique_vals = self.diff().unique()
+            if len(unique_vals) != 1 or unique_vals[0] != self._freq:
+                raise ValueError()
+
+    @_cudf_nvtx_annotate
+    def _copy_type_metadata(
+        self: DatetimeIndex, other: DatetimeIndex, *, override_dtypes=None
+    ) -> GenericIndex:
+        super()._copy_type_metadata(other, override_dtypes=override_dtypes)
+        self._freq = other._freq
+        return self
+
+    @classmethod
+    def _from_data(
+        cls, data: MutableMapping, name: Any = no_default, freq: Any = None
+    ):
+        result = super()._from_data(data, name)
+        result._freq = _validate_freq(freq)
```

Comment:
This validates that `freq` is viable as a `DateOffset`, but not that `freq` is correct for `data`?

Reply:
I added validation in `from_data` too. I haven't made this a method of `DatetimeIndex` because we aren't utilizing any attributes of an existing object(`self`), but don't feel strongly against adding it as a class method. Let me know if either of you feel we need it as class method.